executor.go

```diff
@@ -48,9 +51,160 @@ type ValidationResult struct {
+	// But each validating artifact should be restricted to a single store.
+	// Required.
+	Stores []Store
+
+	// Executor could use multiple verifiers to validate artifacts. Required.
+	Verifiers []Verifier
+
+	// Executor should have at most one policy enforcer to evalute reports. If
+	// not set, the validation result will be returned without evaluation.
+	// Optional.
+	PolicyEnforcer PolicyEnforcer
+}
 
-// ValidateArtifact returns the result of verifying an artifact
+// ValidateArtifact returns the result of verifying an artifact.
 func (e *Executor) ValidateArtifact(ctx context.Context, opts ValidateArtifactOptions) (*ValidationResult, error) {
-	return nil, errors.New("not implemented")
+	aggregatedVerifierReports, err := e.aggregateVerifierReports(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate and aggregate verifier reports: %w", err)
+	}
+
+	if e.PolicyEnforcer == nil {
+		return &ValidationResult{
+			Succeeded:       false,
+			ArtifactReports: aggregatedVerifierReports,
+		}, nil
+	}
+
+	decision, err := e.PolicyEnforcer.Evaluate(ctx, aggregatedVerifierReports)
+	if err != nil {
+		return nil, fmt.Errorf("failed to evaluate verifier reports: %w", err)
+	}
+
+	return &ValidationResult{
+		Succeeded:       decision,
+		ArtifactReports: aggregatedVerifierReports,
+	}, nil
+}
+
+// aggregateVerifierReports generates and aggregates all verifier reports.
+func (e *Executor) aggregateVerifierReports(ctx context.Context, opts ValidateArtifactOptions) ([]*ValidationReport, error) {
+	var aggregatedVerifierReports []*ValidationReport
+	var err error
+
+	// TODO: Implement a worker pool to validate artifacts concurrently.
+	// TODO: Enforce check on the stack size.
+	taskStack := make([]*task, 0)
+
+	// Iterate over all stores to create tasks for each store so that further
+	// referrer artifacts will be fetched from the same store.
+	for _, store := range e.Stores {
+		taskStack = append(taskStack, &task{
+			store:    store,
+			artifact: opts.Subject,
+		})
+	}
+
+	for len(taskStack) > 0 {
+		task := taskStack[len(taskStack)-1]
+		taskStack = taskStack[:len(taskStack)-1]
+
+		taskStack, aggregatedVerifierReports, err = e.processTask(ctx, task, taskStack, aggregatedVerifierReports, opts.ReferenceTypes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate artifact %s: %w", task.artifact, err)
+		}
+	}
+
+	return aggregatedVerifierReports, nil
+}
+
+// processTask processes the task and returns the updated stack and aggregated
+// verifier reports.
+func (e *Executor) processTask(ctx context.Context, task *task, stack []*task, aggregatedVerifierReports []*ValidationReport, referenceTypes []string) ([]*task, []*ValidationReport, error) {
+	artifactRef := task.artifact
+	ref, err := registry.ParseReference(artifactRef)
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to parse artifact reference %s: %w", artifactRef, err)
+	}
+
+	artifactDesc, err := task.store.Resolve(ctx, ref.String())
```

Comment:
Only the root nodes require resolving. Otherwise, it is not required.

We can save out-going calls here.